### PR TITLE
Submission Lynne Maher

### PR DESCRIPTION
**Pull request by LynneTel**

### CODE REVIEW

**Comment by LynneTel** [comment]
Hi,
I know I have taken a long time to do this, but would still appreciate feed back if possible.

Thank you,
Lynne Maher

File: alayatodo/views.py [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
     redirect,
     render_template,
     request,
-    session
-    )
+    session,
+    flash
+)
+import json
+from flask_paginate import Pagination, get_page_args, get_page_parameter
```
Comment (m-a-choquette) [inline alayatodo/views.py]: new dependencies should be added to the requirements.txt.

File: alayatodo/templates/todos.html [hunk 2]
```diff
@@ -14,23 +16,38 @@ <h1>Todo List:</h1>
...
+                        <button type="submit" class="btn btn-xs btn-info"><span class="glyphicon glyphicon-export glyphicon-white"></span></button>
+                    </form>
+                </td>
             </tr>
         {% endfor %}
         <tr>
             <form method="post" action="{{ url_for('todos_POST') }}">
                 <td colspan="3">
-                    <input type="textbox" name="description" class="small-6 small-center" placeholder="Description...">
+                    <input type="textbox" name="description" class="small-6 small-center" placeholder="Description..." required>
```
Comment (m-a-choquette) [inline alayatodo/templates/todos.html]: This should also be validated in the backend.

File: alayatodo/views.py [hunk 3]
```diff
@@ -45,38 +49,120 @@ def logout():
...
+                           page=page,
+                           per_page=per_page,
+                           pagination=pagination)
 
 
 @app.route('/todo', methods=['POST'])
 @app.route('/todo/', methods=['POST'])
 def todos_POST():
+    error = None
+    url = ''
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Those 2 variable are not used.

File: alayatodo/views.py [hunk 4]
```diff
@@ -45,38 +49,120 @@ def logout():
...
-        "INSERT INTO todos (user_id, description) VALUES ('%s', '%s')"
-        % (session['user']['id'], request.form.get('description', ''))
-    )
-    g.db.commit()
-    return redirect('/todo')
+    try:
+        userid = session['user']['id']
+        descr = request.form.get('description', '')
+
+        mytodo = Todo(userid, descr, 0)
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Now I see why the constructor was added to the model, to use positional arguments.
Kwargs are more explicit.

File: alayatodo/views.py [hunk 5]
```diff
@@ -45,38 +49,120 @@ def logout():
...
+        db.session.add(mytodo)
+        db.session.commit()
+        error = "Insert successful!"
```
Comment (m-a-choquette) [inline alayatodo/views.py]: It's a little bit weird to for an `error` variable to hold such a value.

File: alayatodo/views.py [hunk 6]
```diff
@@ -45,38 +49,120 @@ def logout():
...
+    pos = referrer.find('page')
+    url = referrer[pos::]
+
     if not session.get('logged_in'):
         return redirect('/login')
-    g.db.execute("DELETE FROM todos WHERE id ='%s'" % id)
-    g.db.commit()
-    return redirect('/todo')
+    try:
+        mytodo = Todo.query.filter_by(id=id).first()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This allows a user to delete a todo that doesn't belong to another user.

File: alayatodo/views.py [hunk 7]
```diff
@@ -45,38 +49,120 @@ def logout():
...
+@app.route('/todo/', defaults={'page':1}, methods=['GET'])
+@app.route('/todo/page/<int:page>', methods=['GET'])
+@app.route('/todo/page/<int:page>/', methods=['GET'])
+def todos(page):
     if not session.get('logged_in'):
         return redirect('/login')
-    cur = g.db.execute("SELECT * FROM todos")
-    todos = cur.fetchall()
-    return render_template('todos.html', todos=todos)
+    total = Todo.query.count()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This will show all Todos, but it should only show the Todos of the current user.

File: alayatodo/views.py [hunk 8]
```diff
@@ -45,38 +49,120 @@ def logout():
 
 @app.route('/todo/<id>', methods=['GET'])
 def todo(id):
-    cur = g.db.execute("SELECT * FROM todos WHERE id ='%s'" % id)
-    todo = cur.fetchone()
+    todo = Todo.query.filter_by(id=id).first()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This is missing the authentication check.
This allows a user to see a todo that belongs to another user.

File: alayatodo/views.py [hunk 9]
```diff
@@ -45,38 +49,120 @@ def logout():
...
+    url = ''
+    r = request
+    referrer = r.referrer
+
+    pos = referrer.find('page')
+    url = referrer[pos::]
+
+    mytodo = Todo.query.filter_by(id=id).first()
+    with open("data_file.json", "w") as outfile:
+        json.dump({'id':mytodo.id, 'description':mytodo.description}, outfile)
```
Comment (m-a-choquette) [inline alayatodo/views.py]: It's not clear why we need a file to do this.

**Comment by m-a-choquette** [comment]
Hi, 
thanks for submitting your test,
we reviewed it and left some feedback comments.
Have a nice day!

File: alayatodo/views.py [hunk 10]
```diff
@@ -25,11 +31,9 @@ def login_POST():
     username = request.form.get('username')
     password = request.form.get('password')
 
-    sql = "SELECT * FROM users WHERE username = '%s' AND password = '%s'";
-    cur = g.db.execute(sql % (username, password))
-    user = cur.fetchone()
+    user = object_as_dict(User.query.filter_by(username=username).filter_by(password=password).first())
```
[the remark this answers is not in the document]
Comment (m-a-choquette) [reply alayatodo/views.py]: This will crash if the provided credentials are not valid because no user will be returned.

**Pull request closed without merging**